usd/src/usd_parser/utils.cc

```diff
@@ -0,0 +1,282 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "sdf/usd/usd_parser/utils.hh"
+
+#include <pxr/usd/usdShade/input.h>
+#include <pxr/usd/usdShade/material.h>
+#include <pxr/usd/usdShade/shader.h>
+
+#include <ignition/common/Filesystem.hh>
+#include <ignition/common/Util.hh>
+
+#include "sdf/Pbr.hh"
+
+namespace sdf
+{
+  // Inline bracke to help doxygen filtering.
+  inline namespace SDF_VERSION_NAMESPACE {
+  //
+  namespace usd
+  {
+  /////////////////////////////////////////////////
+  std::string removeSubStr(const std::string &_str, const std::string &_substr)
+  {
+    std::string result = _str;
+    size_t pos = std::string::npos;
+    if ((pos = result.find(_substr) )!= std::string::npos)
+    {
+      result.erase(pos, _substr.length());
+    }
+    return result;
+  }
+
+  /////////////////////////////////////////////////
+  bool copyFile(const std::string &_ori, const std::string &_dest)
+  {
+    if (ignition::common::exists(_ori))
+    {
+      std::string baseName = ignition::common::basename(_dest);
+      std::string pathDest = removeSubStr(_dest, baseName);
+      ignition::common::createDirectories(pathDest);
+      return ignition::common::copyFile(_ori, _dest);
+    }
+    return false;
+  }
+
+  /////////////////////////////////////////////////
+  sdf::Material ParseMaterial(const pxr::UsdPrim &_prim)
+  {
+    sdf::Material material;
+    // if the prim is a Geom then get the color values
+    if(_prim.IsA<pxr::UsdGeomGprim>())
+    {
+      auto variant_geom = pxr::UsdGeomGprim(_prim);
+
+      pxr::VtArray<pxr::GfVec3f> color {{0, 0, 0}};
+
+      variant_geom.GetDisplayColorAttr().Get(&color);
+
+      pxr::VtFloatArray displayOpacity;
+      _prim.GetAttribute(
+        pxr::TfToken("primvars:displayOpacity")).Get(&displayOpacity);
+
+      variant_geom.GetDisplayColorAttr().Get(&color);
+      double alpha = 1.0;
+      if (displayOpacity.size() > 0)
+      {
+        alpha = 1 - displayOpacity[0];
+      }
+      material.SetAmbient(
+        ignition::math::Color(
+          ignition::math::clamp(color[0][2] / 0.4, 0.0, 1.0),
+          ignition::math::clamp(color[0][1] / 0.4, 0.0, 1.0),
+          ignition::math::clamp(color[0][0] / 0.4, 0.0, 1.0),
+          alpha));
+      material.SetDiffuse(
+        ignition::math::Color(
+          ignition::math::clamp(color[0][2] / 0.8, 0.0, 1.0),
+          ignition::math::clamp(color[0][1] / 0.8, 0.0, 1.0),
+          ignition::math::clamp(color[0][0] / 0.8, 0.0, 1.0),
+          alpha));
```

Comment:
How is the formula of the colors decided (dividing by 0.4 and 0.8)?

Reply:
I copied this converstion from here https://github.com/ignitionrobotics/sdformat/blob/sdf12/src/parser_urdf.cc#L3181-L3206